// devexpress-dashboard/designer/extensions/options-panel-extension.js
/**
* DevExpress Dashboard (options-panel-extension.js)
* Version:  22.2.4
* Build date: Jan 24, 2023
* Copyright (c) 2012 - 2023 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OptionsPanelExtension = void 0;
const ko = require("knockout");
const control_options_1 = require("../../common/control-options");
const _options_manager_1 = require("../../common/internal/_options-manager");
const _dashboard_item_menu_1 = require("../items/_dashboard-item-menu");
const name = 'item-options-panel';
const nameAlias = 'itemOptionsPanel';
class OptionsPanelExtension {
    constructor(dashboardControl, options = {}) {
        this.dashboardControl = dashboardControl;
        this.name = name;
        this._subscriptions = [];
        this._perMenuSubscriptions = [];
        this._customizeTabsHandlers = [];
        this._optionsManager = new _options_manager_1.OptionsManager();
        this._optionsManager.initialize({
            alias: nameAlias,
            dashboardControl: dashboardControl,
            defaultOptions: {},
            eventsHolder: this,
            initOptions: options,
            optionChanged: (args) => null
        });
    }
    _contextMenuSubscriber(itemContextMenu) {
        if (!!itemContextMenu) {
            var item = this.dashboardControl._actualLayoutController().selectedDashboardItem();
            this._updateContextMenu(itemContextMenu, item, this.dashboardControl.dashboard(), this.dashboardControl._dataSourceBrowser);
        }
    }
    start() {
        var contextMenuExtension = this.dashboardControl.findExtension('item-menu');
        if (contextMenuExtension) {
            this._subscriptions.push(contextMenuExtension._itemContextMenu.subscribe(this._contextMenuSubscriber, this));
            this._contextMenuSubscriber(contextMenuExtension._itemContextMenu());
        }
    }
    stop() {
        this._perMenuSubscriptions.forEach(s => s.dispose());
        this._perMenuSubscriptions = [];
        this._subscriptions.forEach(s => s.dispose());
        this._subscriptions = [];
    }
    _updateContextMenu(menu, item, dashboard, dataSourceBrowser) {
        let composer;
        if (menu.selectedItemSurface) {
            composer = menu.selectedItemSurface.getPropertiesComposer();
            menu.selectedItemSurface.customizeDashboardItemTabs = (args) => {
                this._optionsManager.raiseEvent('customizeSections', args);
            };
        }
        menu.contextMenuItems.push({
            menuItemId: this.name,
            icon: 'dx-dashboard-properties',
            title: 'DashboardWebStringId.Options',
            panelWidth: _dashboard_item_menu_1.DashboardItemMenuSizes.OptionsPanelWidth(),
            templateName: 'dx-dashboard-options',
            detailVisible: ko.observable(false),
            showMenu: () => {
                let args = { dataSourceBrowser, dashboard, propertiesController: menu.propertiesController };
                let tabs = composer.composeTabs(item, args);
                this._customizeTabsHandlers.forEach(handler => handler(tabs, item));
                menu.propertiesController.mainModel({
                    data: {
                        model: item,
                        propertiesTabs: ko.observableArray(tabs),
                    }
                });
            },
            customData: {
                propertiesController: menu.propertiesController,
            },
            index: 300
        });
    }
    _subscribeTabsChanged(handler) {
        if (this._customizeTabsHandlers.indexOf(handler) === -1) {
            this._customizeTabsHandlers.push(handler);
        }
        return {
            dispose: () => {
                this._customizeTabsHandlers.splice(this._customizeTabsHandlers.indexOf(handler), 1);
            }
        };
    }
}
exports.OptionsPanelExtension = OptionsPanelExtension;
control_options_1.designerExtensions[name] = (dashboardControl, options) => new OptionsPanelExtension(dashboardControl, options);
control_options_1.extensionNameMap[name] = nameAlias;
